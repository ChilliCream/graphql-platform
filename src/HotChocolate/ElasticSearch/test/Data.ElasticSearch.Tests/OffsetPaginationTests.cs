using HotChocolate.Data.ElasticSearch.Filters;
using HotChocolate.Execution;
using HotChocolate.Types;
using HotChocolate.Types.Pagination;
using Microsoft.Extensions.DependencyInjection;
using Squadron;
using Xunit;

namespace HotChocolate.Data.ElasticSearch;

[Collection("Elastic Tests")]
public class OffsetPaginationTests : TestBase
{
    private readonly Foo[] _data;

    /// <inheritdoc />
    public OffsetPaginationTests(ElasticsearchResource resource) : base(resource)
    {
        _data = Enumerable
            .Range(0, 20)
            .Select(i => new Foo(i))
            .ToArray();
    }

    [Fact]
    public async Task ElasticSearch_Offset_Pagination_Default_Items()
    {
        await IndexDocuments(_data);

        var executor = await CreateExecutorAsync();

        const string query = """
        query {
            test {
                items {
                    index
                }
                pageInfo {
                    hasNextPage
                    hasPreviousPage
                }
                totalCount
            }
        }
        """;
        var result = await executor.ExecuteAsync(query);
        result.MatchQuerySnapshot();
    }

    [Fact]
    public async Task ElasticSearch_Offset_Pagination_Take_2()
    {
        await IndexDocuments(_data);

        var executor = await CreateExecutorAsync();

        const string query = """
        query {
            test(take: 2) {
                items {
                    index
                }
                pageInfo {
                    hasNextPage
                    hasPreviousPage
                }
                totalCount
            }
        }
        """;

        var result = await executor.ExecuteAsync(query);
        result.MatchQuerySnapshot();
    }

    [Fact]
    public async Task ElasticSearch_Offset_Pagination_Skip_2()
    {
        await IndexDocuments(_data);

        var executor = await CreateExecutorAsync();

        const string query = """
        query {
            test(skip: 2) {
                items {
                    index
                }
                pageInfo {
                    hasNextPage
                    hasPreviousPage
                }
                totalCount
            }
        }
        """;

        var result = await executor.ExecuteAsync(query);
        result.MatchQuerySnapshot();
    }

    [Fact]
    public async Task ElasticSearch_Offset_Pagination_Skip_Take_2()
    {
        await IndexDocuments(_data);

        var executor = await CreateExecutorAsync();

        const string query = """
        query {
            test(skip: 2 take: 2) {
                items {
                    index
                }
                pageInfo {
                    hasNextPage
                    hasPreviousPage
                }
                totalCount
            }
        }
        """;

        var result = await executor.ExecuteAsync(query);
        result.MatchQuerySnapshot();
    }

    private ValueTask<IRequestExecutor> CreateExecutorAsync()
    {
        return new ServiceCollection()
            .AddGraphQL()
            .AddQueryType(x => x
                .Name("Query")
                .Field("test")
                .UseOffsetPaging<ObjectType<Foo>>(options: new PagingOptions { IncludeTotalCount = true })
                .UseTestReport()
                .ResolveTestData<Foo>(Client))
            .AddElasticSearchPagingProvider()
            .BuildTestExecutorAsync();
    }

    public class Foo
    {
        public int Index { get; set; }

        public Foo(int index)
        {
            Index = index;
        }
    }
}
