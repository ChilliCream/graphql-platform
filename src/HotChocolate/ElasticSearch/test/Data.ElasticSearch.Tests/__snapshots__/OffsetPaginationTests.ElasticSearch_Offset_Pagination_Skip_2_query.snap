{
  "from": 2,
  "query": {
    "match_all": {}
  },
  "size": 11
}
